plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/FileSystemExchangeSource.java

```diff
@@ -14,136 +14,293 @@
+                blocked = NOT_BLOCKED;
+            }
+            else {
+                blocked = toCompletableFuture(
                         nonCancellationPropagating(
                                 whenAnyComplete(readers.stream()
                                         .map(ExchangeStorageReader::isBlocked)
-                                        .collect(toImmutableList())))));
+                                        .collect(toImmutableList()))));
             }
-            return this.blocked;
+            blocked = stats.getExchangeSourceBlocked().record(blocked);
+            this.blocked.set(blocked);
+            return blocked;
         }
     }
 
     @Override
     public boolean isFinished()
     {
-        if (closed) {
-            return true;
-        }
-
-        for (ExchangeStorageReader reader : readers) {
-            if (!reader.isFinished()) {
-                return false;
-            }
-        }
-        return true;
+        return closed.get();
     }
 
     @Nullable
     @Override
     public Slice read()
     {
-        if (closed) {
+        if (closed.get()) {
             return null;
         }
 
-        for (ExchangeStorageReader reader : readers) {
+        Slice data = null;
+        List<ExchangeStorageReader> readers = this.readers.get();
+        // regular loop for efficiency
+        for (int i = 0; i < readers.size(); i++) {
+            ExchangeStorageReader reader = readers.get(i);
             if (reader.isBlocked().isDone() && !reader.isFinished()) {
                 try {
-                    return reader.read();
+                    data = reader.read();
+                    break;
                 }
                 catch (IOException e) {
                     throw new UncheckedIOException(e);
                 }
             }
         }
 
-        return null;
+        closeAndCreateReadersIfNecessary();
+
+        return data;
     }
 
     @Override
     public long getMemoryUsage()
     {
         long memoryUsage = 0;
-        for (ExchangeStorageReader reader : readers) {
-            memoryUsage += reader.getRetainedSize();
+        List<ExchangeStorageReader> readers = this.readers.get();
+        // regular loop for efficiency
+        for (int i = 0; i < readers.size(); i++) {
+            memoryUsage += readers.get(i).getRetainedSize();
         }
         return memoryUsage;
     }
 
     @Override
-    public void close()
+    public synchronized void close()
     {
         // Make sure we will only close once
+        if (!closed.compareAndSet(false, true)) {
+            return;
+        }
+        files.clear();
+        Closer closer = Closer.create();
+        for (ExchangeStorageReader reader : readers.getAndSet(ImmutableList.of())) {
+            closer.register(reader);
+        }
+        try {
+            closer.close();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
```

Comment:
Does using Closer make a difference at all?

Reply:
It's safer as it makes sure every object is closed even if one of the `close` method threw and exception.